cmd/gaia/app/export.go

```diff
@@ -0,0 +1,156 @@
+	return appState, validators, nil
+}
+
+// prepare for fresh start at zero height
+func prepForZeroHeightGenesis(ctx sdk.Context) {
+
+	/* TODO XXX check some invariants */
+
+	height := ctx.BlockHeight()
+
+	valAccum := sdk.ZeroDec()
+	vdiIter := func(_ int64, vdi distr.ValidatorDistInfo) bool {
+		lastValPower := app.stakeKeeper.GetLastValidatorPower(ctx, vdi.OperatorAddr)
+		valAccum = valAccum.Add(vdi.GetValAccum(height, sdk.NewDecFromInt(lastValPower)))
+		return false
+	}
+	app.distrKeeper.IterateValidatorDistInfos(ctx, vdiIter)
+
+	lastTotalPower := sdk.NewDecFromInt(app.stakeKeeper.GetLastTotalPower(ctx))
+	totalAccum := app.distrKeeper.GetFeePool(ctx).GetTotalValAccum(height, lastTotalPower)
+
+	if !totalAccum.Equal(valAccum) {
+		panic(fmt.Errorf("validator accum invariance: \n\tfee pool totalAccum: %v"+
+			"\n\tvalidator accum \t%v\n", totalAccum.String(), valAccum.String()))
+	}
+
+	fmt.Printf("accum invariant ok!\n")
+
+	/* END TODO XXX */
+
+	/* Handle fee distribution state. */
+
+	// withdraw all delegator & validator rewards
+	vdiIter = func(_ int64, valInfo distr.ValidatorDistInfo) (stop bool) {
+		err := app.distrKeeper.WithdrawValidatorRewardsAll(ctx, valInfo.OperatorAddr)
+		if err != nil {
+			panic(err)
+		}
+		return false
+	}
+	app.distrKeeper.IterateValidatorDistInfos(ctx, vdiIter)
+
+	ddiIter := func(_ int64, distInfo distr.DelegationDistInfo) (stop bool) {
+		err := app.distrKeeper.WithdrawDelegationReward(
+			ctx, distInfo.DelegatorAddr, distInfo.ValOperatorAddr)
+		if err != nil {
+			panic(err)
+		}
+		return false
+	}
+	app.distrKeeper.IterateDelegationDistInfos(ctx, ddiIter)
+
+	// delete all distribution infos
+	// these will be recreated in InitGenesis
+	app.distrKeeper.RemoveValidatorDistInfos(ctx)
+	app.distrKeeper.RemoveDelegationDistInfos(ctx)
+
+	// assert that the fee pool is empty
+	feePool := app.distrKeeper.GetFeePool(ctx)
+	if !feePool.TotalValAccum.Accum.IsZero() {
+		panic("unexpected leftover validator accum")
+	}
+	bondDenom := app.stakeKeeper.GetParams(ctx).BondDenom
+	if !feePool.ValPool.AmountOf(bondDenom).IsZero() {
+		panic(fmt.Sprintf("unexpected leftover validator pool coins: %v",
+			feePool.ValPool.AmountOf(bondDenom).String()))
+	}
+
+	// reset fee pool height, save fee pool
+	feePool.TotalValAccum.UpdateHeight = 0
+	app.distrKeeper.SetFeePool(ctx, feePool)
+
+	/* Handle stake state. */
+
+	// iterate through validators by power descending, reset bond height, update bond intra-tx counter
+	store := ctx.KVStore(app.keyStake)
+	iter := sdk.KVStoreReversePrefixIterator(store, stake.ValidatorsByPowerIndexKey)
+	counter := int16(0)
+	for ; iter.Valid(); iter.Next() {
+		addr := sdk.ValAddress(iter.Value())
+		validator, found := app.stakeKeeper.GetValidator(ctx, addr)
+		if !found {
+			panic("expected validator, not found")
+		}
+		validator.BondHeight = 0
+		validator.BondIntraTxCounter = counter
+		// AFAICT we do not need to reset unbonding height since it is not used.
```

Comment:
You mentioned in my previous comments that this stuff was removed... can't find the diff for the removed struct fields... also I'm assuming this line is also old and should be removed?

Reply:
I was referring to the height for the delegations, not the unbonding height for the validators. I guess we could remove the latter two.

Reply:
Ah, turns out we do need `validator.UnbondingHeight`, to set the CreationHeights on unbonding delegations and redelegations (which then may later be used to determine whether the delegator should be slashed for an infraction).

Reply:
We now reset `validator.UnbondingHeight` to `0`, which seems fine, since evidence from the old chain won't be valid anyways.

Reply:
👍